packages/engine/Source/DataSources/ModelVisualizer.js

```diff
@@ -445,106 +442,36 @@ ModelVisualizer.prototype.getBoundingSphere = function (entity, result) {
 
-    // Otherwise, in the case of terrain providers with availability,
-    // since the model's bounding sphere may be clamped to a lower LOD tile if
-    // the camera is initially far away, we use sampleTerrainMostDetailed to estimate
-    // where the bounding sphere should be and set that as the target bounding sphere
-    // for the camera.
-    let clampedBoundingSphere = this._modelHash[entity.id]
-      .clampedBoundingSphere;
-
-    // Check if the sample terrain function has failed.
-    const sampleTerrainFailed = this._modelHash[entity.id].sampleTerrainFailed;
-    if (sampleTerrainFailed) {
-      this._modelHash[entity.id].sampleTerrainFailed = false;
-      return BoundingSphereState.FAILED;
+    // Regardless of what the original model's position is set to, when clamping we reset it to 0
+    // when computing the position to zoom/fly to.
+    if (isHeightReferenceClamp(model.heightReference)) {
+      cartoPosition.height = 0;
     }
 
-    if (!defined(clampedBoundingSphere)) {
-      clampedBoundingSphere = new BoundingSphere();
-
-      // Since this function is called per-frame, we set a flag when sampleTerrainMostDetailed
-      // is called and check for it to avoid calling it again.
-      const awaitingSampleTerrain = this._modelHash[entity.id]
-        .awaitingSampleTerrain;
-      if (!awaitingSampleTerrain) {
-        Cartographic.clone(cartoPosition, scratchCartographic);
-        this._modelHash[entity.id].awaitingSampleTerrain = true;
-        ModelVisualizer._sampleTerrainMostDetailed(terrainProvider, [
-          scratchCartographic,
-        ])
-          .then((result) => {
-            if (this.isDestroyed()) {
-              return;
-            }
-
-            this._modelHash[entity.id].awaitingSampleTerrain = false;
-
-            const updatedCartographic = result[0];
-            if (model.heightReference === HeightReference.RELATIVE_TO_GROUND) {
-              updatedCartographic.height += cartoPosition.height;
-            }
-            ellipsoid.cartographicToCartesian(
-              updatedCartographic,
-              scratchPosition
-            );
-
-            // Update the bounding sphere with the updated position.
-            BoundingSphere.clone(model.boundingSphere, clampedBoundingSphere);
-            clampedBoundingSphere.center = scratchPosition;
-
-            this._modelHash[
-              entity.id
-            ].clampedBoundingSphere = BoundingSphere.clone(
-              clampedBoundingSphere
-            );
-          })
-          .catch((e) => {
-            if (this.isDestroyed()) {
-              return;
-            }
-
-            this._modelHash[entity.id].sampleTerrainFailed = true;
-            this._modelHash[entity.id].awaitingSampleTerrain = false;
-          });
+    const height = scene.getHeight(cartoPosition, model.heightReference);
+    if (defined(height)) {
+      if (isHeightReferenceClamp(model.heightReference)) {
+        cartoPosition.height = height;
```

Comment:
The fact that this check is happening twice and the assignment is happening twice for this code path feels wrong to me. Is there a way to rewrite this to be more clear what value the height should be in each scenario or defined/undefined reference and clamp or not?